codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -69,60 +69,113 @@
+            this.validateHttp2Headers = validate;
+            return (B) this;
         }
-    }
 
-    public Http2ConnectionHandler(Http2Connection connection, Http2Settings initialSettings) {
-        this(connection, new DefaultHttp2FrameReader(), new DefaultHttp2FrameWriter(),
-                initialSettings);
-    }
+        @SuppressWarnings("unchecked")
+        public B initialSettings(Http2Settings settings) {
+            initialSettings = settings;
+            return (B) this;
+        }
 
-    public Http2ConnectionHandler(Http2Connection connection, Http2FrameReader frameReader,
-                                  Http2FrameWriter frameWriter, Http2Settings initialSettings) {
-        this.initialSettings = initialSettings;
-        encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
-        decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, frameReader);
+        public final Http2Settings initialSettings() {
+            return initialSettings;
+        }
+
+        @SuppressWarnings("unchecked")
+        public B server(boolean isServer) {
+            server = isServer;
+            return (B) this;
+        }
+
+        @SuppressWarnings("unchecked")
+        public B frameLogger(Http2FrameLogger logger) {
+            this.frameLogger = logger;
+            return (B) this;
+        }
+
+        @SuppressWarnings("unchecked")
+        public B gracefulShutdownTimeoutMillis(long gracefulShutdownTimeoutMillis) {
+            this.gracefulShutdownTimeoutMillis = gracefulShutdownTimeoutMillis;
+            return (B) this;
+        }
+
+        public long gracefulShutdownTimeoutMillis() {
+            return gracefulShutdownTimeoutMillis;
+        }
+
+        public final T build() {
+            return build(new DefaultHttp2Connection(server));
+        }
+
+        public final T build(Http2Connection connection) {
+            Http2FrameReader reader = new DefaultHttp2FrameReader(validateHttp2Headers);
+            Http2FrameWriter writer = new DefaultHttp2FrameWriter();
+            Http2FrameLogger logger = frameLogger;
+            if (logger != null) {
+                reader = new Http2InboundFrameLogger(reader, logger);
+                writer = new Http2OutboundFrameLogger(writer, logger);
+            }
+            return build(connection, reader, writer);
+        }
+
+        public final T build(Http2Connection connection, Http2FrameReader frameReader, Http2FrameWriter frameWriter) {
+            Http2ConnectionEncoder encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
+            Http2ConnectionDecoder decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, frameReader);
+            return build(decoder, encoder);
+        }
+
+        public abstract T build(Http2ConnectionDecoder decoder, Http2ConnectionEncoder encoder);
+
+        protected void setPostBuildOptions(Http2ConnectionHandler handler) {
```

Comment:
Maybe get rid of the frameListener from the base class and just set it from the concrete class if it's appropriate?